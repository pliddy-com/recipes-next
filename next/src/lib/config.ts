interface PageProps {
  defaultDescription?: string;
  defaultTitle?: string;
  description?: string;
  domain?: string;
  title?: string;
}

interface MicrocopyProps {
  notFound: PageProps;
  index: PageProps;
  recipe: PageProps;
  site: PageProps;
  tag: PageProps;
  tagIndex: PageProps;
}

export interface ConfigProps {
  microcopy?: MicrocopyProps;
}

const microcopy: MicrocopyProps = {
  index: {
    defaultTitle: 'All the Recipes',
    description: "Patrick's recipe collection"
  },
  notFound: {
    defaultTitle: "Sorry, that page isn't here.",
    description: 'The requested url was not found'
  },
  recipe: {
    defaultDescription: 'A recipe for',
    defaultTitle: 'Recipe'
  },
  site: {
    title: "Patrick's Recipes",
    domain: 'https://recipes.pliddy.com'
  },
  tag: {
    defaultDescription: 'A collection of recipies tagged as',
    defaultTitle: 'Collection'
  },
  tagIndex: {
    defaultTitle: 'Tag Collections',
    description: 'All tag collections that contain recipes.'
  }
};

const config: ConfigProps = {
  microcopy
};

export default config;
