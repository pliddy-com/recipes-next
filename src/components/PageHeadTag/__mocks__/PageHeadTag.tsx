// Mock modules for npm packages are in the __mocks__ directory
// at the root level as a sibling of node_modules
import Head from 'next/head';
import { ImageDefaultFragment } from 'types/queries';

// create mock version of npm package
jest.genMockFromModule('../PageHeadTag');

jest.mock('next/head');

interface PageHeadTagProps {
  title?: string | null;
  defaultTitle?: string;
  description?: string;
  image?: ImageDefaultFragment | null;
}

export const PageHeadTag = jest.fn(
  ({ title, defaultTitle, description, image }: PageHeadTagProps) => {
    return (
      <Head>
        <title>{title || defaultTitle}</title>
        <meta name="description" content={description} />

        {image && (
          <meta property="og:image" content={`${image.url}?w=400&h=300`} />
        )}
      </Head>
    );
  }
);

export default PageHeadTag;
