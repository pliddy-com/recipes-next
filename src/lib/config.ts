interface PageProps {
  defaultDescription?: string;
  defaultTitle?: string;
  description?: string;
  domain?: string;
  title?: string;
}

interface MicrocopyProps {
  category?: PageProps;
  index?: PageProps;
  recipe?: PageProps;
  site?: PageProps;
  tag?: PageProps;
}

export interface ConfigProps {
  microcopy?: MicrocopyProps;
}

const microcopy: MicrocopyProps = {
  category: {
    defaultDescription: 'A collection of recipies categorized as',
    defaultTitle: 'Category',
  },
  index: {
    defaultTitle: 'All Recipes',
    description: "Patrick's recipe collection",
  },
  recipe: {
    defaultDescription: 'A recipe for',
    defaultTitle: 'Recipe',
  },
  site: {
    title: "Patrick's Recipes",
    domain: 'https://recipes.pliddy.com',
  },
  tag: {
    defaultDescription: 'A collection of recipies tagged as',
    defaultTitle: 'Tag',
  },
};

const config: ConfigProps = {
  microcopy,
};

export default config;
