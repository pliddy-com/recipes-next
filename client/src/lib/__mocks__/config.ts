import { ConfigProps } from 'lib/config';

const microcopy = {
  index: {
    defaultTitle: 'Index Default Title',
    description: 'Index default description.'
  },
  notFound: {
    defaultTitle: 'Page Not Found',
    description: 'The requested url was not found'
  },
  recipe: {
    defaultDescription: 'Recipe default description for',
    defaultTitle: 'Recipe Default Title'
  },
  search: {
    defaultTitle: 'Results for',
    description: 'Search results'
  },
  site: {
    title: 'Site Title',
    domain: 'https://recipes.pliddy.com'
  },
  tag: {
    defaultDescription: 'Tag default description for',
    defaultTitle: 'Tag Default Title'
  },
  tagIndex: {
    defaultTitle: 'Tag Index Default Title',
    description: 'Tag index default description.'
  }
};

const config: ConfigProps = {
  microcopy
};

module.exports = config;
