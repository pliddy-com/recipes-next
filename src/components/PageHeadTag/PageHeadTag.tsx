import Head from 'next/head';

import config from 'lib/config';
import { useRouter } from 'next/router';
import { ImageDefaultFragment } from 'types/generated/graphql';

interface PageHeadTagProps {
  title?: string | null;
  defaultTitle?: string;
  description?: string;
  image?: ImageDefaultFragment | null;
}

const PageHeadTag = ({
  title,
  defaultTitle,
  description,
  image,
}: PageHeadTagProps) => {
  const { asPath } = useRouter();
  const { site, index } = config?.microcopy ?? {};
  const { title: siteTitle, domain } = site ?? {};
  const { description: defaultDescription } = index ?? {};

  const imgWidth = 800;
  const imgHeight = 320;
  const locale = 'en_US';
  const tagDescription = description || defaultDescription;
  const tagTitle = title || defaultTitle;
  const type = image ? 'article' : 'website';
  const url = `${domain}${asPath}`;

  const canonicalPath = asPath.includes('category')
    ? `${asPath.replace('category', 'tag')}`
    : `${asPath}`;

  const canonicalUrl = `${domain}${canonicalPath}`;

  return siteTitle || defaultTitle || description || defaultDescription ? (
    <Head>
      {siteTitle && tagTitle && <title>{`${siteTitle} | ${tagTitle}`}</title>}
      {tagDescription && <meta name="description" content={tagDescription} />}

      <meta property="og:description" content={tagDescription} />

      {image && image.url && (
        <meta
          property="og:image"
          content={`${image.url}?w=${imgWidth}&h=${imgHeight}`}
        />
      )}
      {image && image.description && (
        <meta property="og:image:alt" content={image.description} />
      )}
      {image && <meta property="og:image:height" content={`${imgHeight}`} />}
      {image && <meta property="og:image:width" content={`${imgWidth}`} />}
      {image && <meta property="og:image:type" content="image/jpeg" />}

      <meta property="og:locale" content={locale} />
      <meta property="og:site_name" content={siteTitle} />
      <meta property="og:title" content={tagTitle} />
      <meta property="og:type" content={type} />
      <meta property="og:url" content={url} />

      {canonicalUrl && <link rel="canonical" href={canonicalUrl} />}
    </Head>
  ) : null;
};

export default PageHeadTag;
