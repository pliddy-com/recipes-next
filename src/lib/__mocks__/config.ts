import { ConfigProps } from 'lib/config';

const microcopy = {
  category: {
    defaultDescription: 'Category default description',
    defaultTitle: 'Category Default Title',
  },
  index: {
    defaultTitle: 'Index Default Title',
    description: 'Index default description.',
  },
  recipe: {
    defaultDescription: 'Recipe default description',
    defaultTitle: 'Recipe Default Title',
  },
  site: {
    title: 'Site Title',
  },
  tag: {
    defaultDescription: 'Tag default description',
    defaultTitle: 'Tag Default Title',
  },
};

const config: ConfigProps = {
  microcopy,
};

module.exports = config;
