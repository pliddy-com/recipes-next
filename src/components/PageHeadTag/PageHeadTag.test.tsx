import { render } from '@testing-library/react';

// add custom jest matchers from jest-dom
import '@testing-library/jest-dom';

// import the component to test
import PageHeadTag from './PageHeadTag';

import config from 'lib/config';

jest.mock('lib/config');
jest.mock('next/head');

describe('PageHeadingTag', () => {
  // reset mocks after each test
  afterEach(() => {
    jest.resetModules();
  });

  describe('when there is a config object', () => {
    it('it renders a Next Head tag with a title and description', () => {
      const title = 'Test Title';
      const description = 'Decription';

      const { container } = render(
        <PageHeadTag title={title} description={description} />
      );

      const titleTag = document.getElementsByTagName('title')[0];

      expect(titleTag).toBeInTheDocument();
      expect(titleTag.text.includes(title));

      const descriptionTag = document.querySelector(
        `meta[content="${description}"]`
      );

      expect(descriptionTag).toBeInTheDocument();

      // assert that the component matches the existing snapshot
      expect(container).toMatchSnapshot();
    });

    it('it renders a Next Head tag with a default title and description', () => {
      const defaultTitle = config?.microcopy?.site?.title;
      const defaultDescription = config?.microcopy?.index?.description;

      render(<PageHeadTag defaultTitle={defaultTitle} />);

      const titleTag = document.getElementsByTagName('title')[0];

      expect(titleTag).toBeInTheDocument();
      defaultTitle && expect(titleTag.text.includes(defaultTitle));

      const descriptionTag = document.querySelector(
        `meta[content="${defaultDescription}"]`
      );

      expect(descriptionTag).toBeInTheDocument();
    });

    it('it renders a Next Head tag if no values defaults are passed', () => {
      const defaultDescription = config?.microcopy?.index?.description;

      render(<PageHeadTag />);

      const titleTag = document.getElementsByTagName('title')[0];
      const descriptionTag = document.querySelector(
        `meta[content="${defaultDescription}"]`
      );

      expect(titleTag).toBeUndefined();
      expect(descriptionTag).toBeInTheDocument();
    });
  });

  describe('when there is no config object', () => {
    // before each test, delete microcopy node from config
    beforeEach(() => {
      delete config.microcopy;
    });

    it('it does not render', () => {
      // no tags props, so should return null
      render(<PageHeadTag />);

      const titleTag = document.getElementsByTagName('title')[0];
      expect(titleTag).toBeUndefined();
    });
  });
});
