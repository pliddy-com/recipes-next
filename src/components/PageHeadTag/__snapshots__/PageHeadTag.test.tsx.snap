// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHeadingTag when there is a config object it renders a Next Head tag with a title and description 1`] = `
<div>
  <title>
    Site Title | Test Title
  </title>
  <meta
    content="Decription"
    name="description"
  />
  <meta
    content="Decription"
    property="og:description"
  />
  <meta
    content="https://image.url?w=750&h=300"
    property="og:image"
  />
  <meta
    content="Image Description"
    property="og:image:alt"
  />
  <meta
    content="300"
    property="og:image:height"
  />
  <meta
    content="750"
    property="og:image:width"
  />
  <meta
    content="image/jpeg"
    property="og:image:type"
  />
  <meta
    content="en_US"
    property="og:locale"
  />
  <meta
    content="Site Title"
    property="og:site_name"
  />
  <meta
    content="Test Title"
    property="og:title"
  />
  <meta
    content="article"
    property="og:type"
  />
  <meta
    content="https://recipes.pliddy.com/category/path"
    property="og:url"
  />
  <link
    href="https://recipes.pliddy.com/tag/path"
    rel="canonical"
  />
</div>
`;
